File: src/game/world/position.ts
export interface Position {
  x: number;
  y: number;
}

export function pos(x: number, y: number): Position {
  return { x, y };
}

export function posKey(position: Position): string {
  return `${position.x},${position.y}`;
}

export function parsePosition(text: string): Position {
  const [x, y] = text.split(",").map((part) => Number.parseInt(part.trim(), 10));
  return { x, y };
}

export function samePosition(a: Position, b: Position): boolean {
  return a.x === b.x && a.y === b.y;
}

/** Chebyshev distance, which is how many diagonal steps apart two tiles are. */
export function distance(a: Position, b: Position): number {
  return Math.max(Math.abs(a.x - b.x), Math.abs(a.y - b.y));
}

export function translate(position: Position, dx: number, dy: number): Position {
  return { x: position.x + dx, y: position.y + dy };
}

File: game/src/main/java/com/exjets/aetheria/world/Position.java
package com.exjets.aetheria.world;

/** An immutable tile coordinate. */
public record Position(int x, int y) {

    public int distanceTo(Position other) {
        return Math.max(Math.abs(x - other.x), Math.abs(y - other.y));
    }

    public boolean isAdjacentTo(Position other) {
        return distanceTo(other) <= 1;
    }

    /** True when the tiles share an edge, which is what interactions require. */
    public boolean isOrthogonallyAdjacentTo(Position other) {
        return Math.abs(x - other.x) + Math.abs(y - other.y) == 1;
    }

    public Position translate(int dx, int dy) {
        return new Position(x + dx, y + dy);
    }

    @Override
    public String toString() {
        return x + "," + y;
    }

    public static Position parse(String text) {
        String[] parts = text.split(",");
        return new Position(Integer.parseInt(parts[0].trim()), Integer.parseInt(parts[1].trim()));
    }
}
